pkg/pop/poptoken.go

```diff
@@ -0,0 +1,159 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+package pop
+
+import (
+	"crypto"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha256"
+	"encoding/base64"
+	"fmt"
+	"math/big"
+	"sync"
+)
+
+// PoPKey is a generic interface for PoP key properties and methods
+type PoPKey interface {
+	// encryption/signature algo
+	Alg() string
+	// kid
+	KeyID() string
+	// jwk that can be embedded in JWT w/ PoP token's cnf claim
+	JWK() string
+	// https://tools.ietf.org/html/rfc7638 compliant jwk thumbprint
+	JWKThumbprint() string
+	// req_cnf claim that can be included in access token request to AAD
+	ReqCnf() string
+	// sign payload using private key
+	Sign([]byte) ([]byte, error)
+}
+
+// software based pop key implementation of PoPKey
+type swKey struct {
+	key    *rsa.PrivateKey
+	keyID  string
+	jwk    string
+	jwkTP  string
+	reqCnf string
+}
+
+// Alg returns the algorithm used to encrypt/sign the swKey
+func (swk *swKey) Alg() string {
+	return "RS256"
+}
+
+// KeyID returns the keyID of the swKey, representing the key used to sign the swKey
+func (swk *swKey) KeyID() string {
+	return swk.keyID
+}
+
+// JWK returns the JSON Web Key of the given swKey
+func (swk *swKey) JWK() string {
+	return swk.jwk
+}
+
+// JWKThumbprint returns the JWK thumbprint of the given swKey
+func (swk *swKey) JWKThumbprint() string {
+	return swk.jwkTP
+}
+
+// ReqCnf returns the req_cnf claim to send to AAD for the given swKey
+func (swk *swKey) ReqCnf() string {
+	return swk.reqCnf
+}
+
+// Sign uses the given swKey to sign the given payload and returns the signed payload
+func (swk *swKey) Sign(payload []byte) ([]byte, error) {
+	return swk.key.Sign(rand.Reader, payload, crypto.SHA256)
+}
+
+// init initializes the given swKey using the given private key
+func (swk *swKey) init(key *rsa.PrivateKey) {
+	swk.key = key
+
+	eB64, nB64 := getRSAKeyExponentAndModulus(key)
+	swk.jwkTP = computeJWKThumbprint(eB64, nB64)
+	swk.reqCnf = getReqCnf(swk.jwkTP)
+
+	// set keyID to jwkTP
+	swk.keyID = swk.jwkTP
+
+	// compute JWK to be included in JWT w/ PoP token's cnf claim
+	// - https://tools.ietf.org/html/rfc7800#section-3.2
+	swk.jwk = getJWK(eB64, nB64, swk.keyID)
+}
+
+// generateSwKey generates a new swkey and initializes it with required fields before returning it
+func generateSwKey(key *rsa.PrivateKey) (*swKey, error) {
+	swk := &swKey{}
+	swk.init(key)
+	return swk, nil
+}
+
+var pswKey *swKey
+var pwsKeyMutex sync.Mutex
+
+// GetSwPoPKey generates a new PoP key that rotates every 8 hours and returns it
+func GetSwPoPKey() (*swKey, error) {
+	key, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return nil, fmt.Errorf("error generating RSA private key: %w", err)
+	}
+	return GetSwPoPKeyWithRSAKey(key)
+}
+
+func GetSwPoPKeyWithRSAKey(rsaKey *rsa.PrivateKey) (*swKey, error) {
+	pwsKeyMutex.Lock()
```

Comment:
the lock `pwsKeyMutex` and the global key `pswKey` can be removed since you are not using this anymore.

this funciton can just be 

```
	key, err := generateSwKey(rsaKey)
	if err != nil {
		return nil, fmt.Errorf("unable to generate popkey. err: %w", err)
	}
	return key, nil
```

Reply:
Cleaned up the function